cmd/istiod/istiod.go

```diff
@@ -0,0 +1,124 @@
+// - includes a Secret controller that provisions certs as secrets.
+//
+// Normal hyperistio is using local config files and MCP sources for config/endpoints,
+// as well as SDS backed by a file-based CA.
+func main() {
+	stop := make(chan struct{})
+
+	// First create the k8s clientset - and return the config source.
+	// The config includes the address of apiserver and the public key - which will be used
+	// after cert generation, to check that Apiserver-generated certs have same key.
+	client, kcfg, err := k8s.CreateClientset(os.Getenv("KUBECONFIG"), "")
+	if err != nil {
+		log.Fatal("Failed to connect to k8s", err)
+	}
+
+	// Load the mesh config. Note that the path is slightly changed - attempting to move all istio
+	// related under /var/lib/istio, which is also the home dir of the istio user.
+	istiods, err := istiod.InitConfig("/var/lib/istio/config")
+	if err != nil {
+		log.Fatal("Failed to start istiod ", err)
+	}
+
+	// Create k8s-signed certificates. This allows injector, validation to work without Citadel, and
+	// allows secure SDS connections to Istiod.
+	initCerts(istiods, client, kcfg)
+
+	// Init k8s related components, including Galley K8S controllers and
+	// Pilot discovery. Code kept in separate package.
+	k8sServer, err := k8s.InitK8S(istiods, client, kcfg, istiods.Args)
+	if err != nil {
+		log.Fatal("Failed to start k8s controllers ", err)
+	}
+
+	// Initialize Galley config source for K8S.
+	galleyK8S, err := k8sServer.NewGalleyK8SSource(istiods.Galley.Resources)
+	istiods.Galley.Sources = append(istiods.Galley.Sources, galleyK8S)
+
+	err = istiods.InitDiscovery()
+	if err != nil {
+		log.Fatal("Failed to init XDS server ", err)
+	}
+
+	k8sServer.InitK8SDiscovery(istiods, client, kcfg, istiods.Args)
+
+
+	err = istiods.Start(stop, k8sServer.OnXDSStart)
+	if err != nil {
+		log.Fatal("Failure on start XDS server", err)
+	}
+
+	k8sServer.StartSDSK8S(istiods.Mesh)
+
+	// Feedback: don't run envoy sidecar in istiod
+	//err = s.StartEnvoy(baseDir, kc.IstioServer.Mesh)
+	//if err != nil {
+	//	log.Fatal("Failure on start istio-control sidecar", err)
+	//}
+
+	// Injector should run along, even if not used - but only if the injection template is mounted.
+	if _, err := os.Stat("./var/lib/istio/inject/injection-template.yaml"); err == nil {
+		err = k8s.StartInjector(stop)
+		if err != nil {
+			log.Fatalf("Failure to start injector ", err)
+		}
+	}
+
+	istiods.Serve(stop)
+	istiods.WaitStop(stop)
+}
+
+func initCerts(server *istiod.Server, client *kubernetes.Clientset, cfg *rest.Config) {
+	// TODO: fallback to citadel (or custom CA)
+
+	certChain, keyPEM, err := k8s.GenKeyCertK8sCA(client.CertificatesV1beta1(), "istio-system",
+		"istio-pilot.istio-system,istiod.istio-system")
+	if err != nil {
+		log.Fatal("Failed to initialize certs")
+	}
+	server.CertChain = certChain
+	server.CertKey = keyPEM
+
+	// Save the certificates to /var/run/secrets/istio-dns
+	os.MkdirAll(istiod.DnsCertDir, 0700)
```

Comment:
this may fail on k8s due to permissions and we don't handle the error

Reply:
I tested it in k8s - seems to work, but I can look for an alternative.

Reply:
It didn't work for me but maybe I had something wrong in my setup

Reply:
I updated the k8s file to mount an empty dir.